stdlib/public/core/FloatingPointTypes.swift.gyb

```diff
@@ -1599,17 +1608,26 @@ extension ${Self} {
   /// can't be represented exactly, the result is `nil`.
   ///
   /// - Parameter value: The integer to represent as a floating-point value.
-%   if srcBits < SignificandBitCount:
+%   if srcBits <= dstBits:
   @available(*, message: "Converting ${That} to ${Self} will always succeed.")
 %   end
   @inlinable // FIXME(sil-serialize-all)
   @inline(__always)
   public init?(exactly value: ${That}) {
     _value = Builtin.${sign}itofp_${ThatBuiltinName}_FPIEEE${bits}(value._value)
 
-%   if srcBits >= SignificandBitCount:
-    guard let roundTrip = ${That}(exactly: self),
-      roundTrip == value else {
+%   if srcBits > dstBits:
+    // The magnitude of the integer must be representable by the significand.
+    // Omit trailing zeros that can be encoded by the exponent.
+    let absv = value.magnitude
+    let mask: ${That}.Magnitude = ~((1 << (${SignificandBitCount} + 1)) - 1)
+    let tz = absv.trailingZeroBitCount
+    if _slowPath((absv >> tz) & mask != 0) {
+      return nil
+    }
+    // Check that the required exponent is in range.
+    let maxexp = (1 << (${ExponentBitCount} - 1)) - 1
+    if _slowPath(absv >> maxexp > 1) {
```

Comment:
This isn't quite right, I don't think. If a value has fewer ones than `SignificandBitCount`, but exactly the number of trailing zeros so that `absv >> maxexp == 1`, then this algorithm will return an "exact" value that's inexact.

Reply:
I'm not sure what you mean here. `absv` hasn't been shifted by the number of trailing zeros, so this check should be enough to check that it is in range.

For example the largest finite value a 32-bit Float can represent is `0xffffff << 104`. `(0xffffff << 104) >> 127 == 1` and so passes the check. Higher values (e.g. `0x1000000 << 104`) are equal to 2 or more when shifted right by 127 *or* do not have enough trailing zeros and will be caught by the earlier check.

That said I do think we need to hoist this out of its current position so we can do the exponent check separately.